src/meta/src/manager/scale.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+use std::sync::atomic::{AtomicU32, Ordering};
+use std::sync::Arc;
+
+use log::error;
+use risingwave_common::error::{ErrorCode, Result};
+use risingwave_pb::meta::scale_task::TaskType;
+use risingwave_pb::meta::{ScaleTask, TaskStatus};
+use tokio::sync::{Mutex, Semaphore};
+
+use crate::model::{MetadataModel, ScaleTaskId};
+use crate::storage::MetaStore;
+
+pub type ScaleManagerRef<S> = Arc<ScaleManager<S>>;
+
+/// [`ScaleManager`] is used to receive scale tasks.
+#[allow(dead_code)]
+pub struct ScaleManager<S: MetaStore> {
+    meta_store: Arc<S>,
+
+    next_task_id: AtomicU32,
+    task_queue: Arc<Mutex<VecDeque<ScaleTask>>>,
+    task_semaphore: Arc<Semaphore>,
+}
+
+impl<S> ScaleManager<S>
+where
+    S: MetaStore,
+{
+    pub fn new(meta_store: Arc<S>) -> Self {
+        // TODO(zehua): read task from meta_store.
+        let task_queue = Arc::new(Mutex::new(VecDeque::new()));
+        let task_semaphore = Arc::new(Semaphore::new(0));
+        let scale_manager = ScaleManager {
+            meta_store: meta_store.clone(),
+            next_task_id: AtomicU32::new(1),
+            task_queue: task_queue.clone(),
+            task_semaphore: task_semaphore.clone(),
+        };
+
+        tokio::spawn(async move {
```

Comment:
Please add a shutdown channel and expose the join handle, so that we can graceful shutdown scale task handling.